File: app/src/main/java/ms/imagine/foodiemate/adapter/EggsDetailAdapter.java
package ms.imagine.foodiemate.adapter;

import android.content.Context;
import android.support.v4.content.ContextCompat;
import android.support.v7.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageView;
import android.widget.TextView;
import ms.imagine.foodiemate.R;
import ms.imagine.foodiemate.data.Egg;
import ms.imagine.foodiemate.data.Eggs;
import ms.imagine.foodiemate.presenters.FbStorageRead;


/**
 * Created by eugen on 3/30/2018.
 */
public class EggsDetailAdapter extends RecyclerView.Adapter<RecyclerView.ViewHolder> {
    private Eggs myDataset;
    private Context context;
    private FbStorageRead fbStorageRead;
    private OnItemClicked onClick;

    public EggsDetailAdapter(Eggs myDataset, Context context) {
        this.myDataset = myDataset;
        this.context = context;
        this.fbStorageRead = new FbStorageRead();
    }


    static class VH_HEAD extends RecyclerView.ViewHolder {
        ImageView imgView;
        TextView status;

        VH_HEAD(View itemView) {
            super(itemView);
            imgView = itemView.findViewById(R.id.imgview_thumbnail);
            status = itemView.findViewById(R.id.egg_info);
        }
    }


    static class VH_BODY extends RecyclerView.ViewHolder {
        TextView status;
        ImageView imgView;

        VH_BODY(View view) {
            super(view);
            status = view.findViewById(R.id.status);
            imgView = view.findViewById(R.id.imgView);
        }
    }

    @Override
    public RecyclerView.ViewHolder onCreateViewHolder(ViewGroup parent, int viewType) {
        if (viewType == 0) {
            return new VH_HEAD(LayoutInflater.from(parent.getContext())
                    .inflate(R.layout.card_detailhead, parent, false));
        } else {
            return new VH_BODY(LayoutInflater.from(parent.getContext())
                    .inflate(R.layout.card_detailbody, parent, false));
        }
    }

    @Override
    public void onBindViewHolder(RecyclerView.ViewHolder raw_holder, int position) {
        Eggs egg = myDataset;
        if (raw_holder instanceof VH_HEAD) {
            VH_HEAD holder = (VH_HEAD) raw_holder;
            holder.status.setText(egg.status());
            holder.imgView.setImageResource(egg.getThumbnail());
        }
        if (raw_holder instanceof VH_BODY) {
            VH_BODY holder = (VH_BODY) raw_holder;
            if (egg.isLegacyEgg()) {
                holder.status.setText(egg.status());
                if (egg.isIsnewEgg()) {
                    holder.imgView.setImageURI(egg.getLocalImgUri());
                } else {
                    fbStorageRead.downloadImage(context, egg.getRemoteImgURL(), holder.imgView);
                }
            } else {
                Egg place = (egg.getEgg().get(position - 1));
                holder.status.setText(place.getStatus());
                holder.imgView.setImageURI(egg.getLocalImgUri());
                fbStorageRead.downloadImage(context, egg.getEgg().get(position - 1).getRemoteImgURL(), holder.imgView);
            }

        }
    }


    @Override
    public int getItemCount() {
        if (myDataset.isLegacyEgg()) {
            return 2;
        } else {
            return myDataset.getEgg().size() + 1;
        }
    }


    interface OnItemClicked {
        void onItemClick(int position);
    }

    @Override
    public int getItemViewType(int position) {
        if (position == 0) {
            return 0;
        } else {
            return 1;
        }
    }
}

File: app/src/main/java/ms/imagine/foodiemate/adapter/EggsDetailAdapter.kt
package ms.imagine.foodiemate.adapter

import android.content.Context
import android.support.v7.widget.RecyclerView
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.ImageView
import android.widget.TextView
import ms.imagine.foodiemate.Presenter.FbStorageRead
import ms.imagine.foodiemate.R
import ms.imagine.foodiemate.data.Eggs


/**
 * Created by eugen on 3/30/2018.
 */
class EggsDetailAdapter(private val myDataset: Eggs, private val context: Context) :
        RecyclerView.Adapter<RecyclerView.ViewHolder>() {
    val fbStorageRead =  FbStorageRead()

    inner class VH_HEAD(view: View) : RecyclerView.ViewHolder(view) {
        val imgView: ImageView = view.findViewById<View>(R.id.imgview_thumbnail) as ImageView
        val status: TextView = view.findViewById<View>(R.id.egg_info) as TextView
    }


    inner class VH_BODY(view: View) : RecyclerView.ViewHolder(view) {
        val status: TextView = view.findViewById<View>(R.id.status) as TextView
        val imgView: ImageView = view.findViewById<View>(R.id.imgView) as ImageView
    }

    override fun onBindViewHolder(holder: RecyclerView.ViewHolder, position: Int) {
        val egg = myDataset
        if(holder is VH_HEAD){
            holder.status.text = egg.displayStatus()
            holder.imgView.setImageDrawable(egg.displayStatusThumbnail(context))
        }


        if (holder is VH_BODY){
            if (egg.isLegacyEgg()){
                holder.status.text = egg.displayStatus()
                if (egg.isnewEgg) {
                    holder.imgView.setImageURI(egg.localImgUri)
                } else {
                    fbStorageRead.downloadImage(context, egg.remoteImgURL, holder.imgView)
                }
            } else {
                holder.status.text = egg.egg.get(position-1).status.toString()
                holder.imgView.setImageURI(egg.localImgUri)
                fbStorageRead.downloadImage(context, egg.egg.get(position-1).remoteImgURL, holder.imgView)
            }

        }
        //holder.entity.setOnClickListener { onClick.onItemClick(position) }
    }

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): RecyclerView.ViewHolder {
        return if (viewType == 0) {
            VH_HEAD(LayoutInflater.from(parent.context)
                    .inflate(R.layout.card_detailhead, parent, false) as View)
        } else {
            VH_BODY(LayoutInflater.from(parent.context)
                    .inflate(R.layout.card_detailbody, parent, false) as View)
        }
    }

    override fun getItemCount(): Int {
        return if(myDataset.isLegacyEgg()){
            2
        } else {
            myDataset.egg.size + 1
        }
    }




    private lateinit var onClick: OnItemClicked


    interface OnItemClicked {
        fun onItemClick(position: Int)
    }

    override fun getItemViewType(position: Int): Int {
        return if(position ==0) {
            0
        } else {
            1
        }
    }

    companion object {
        val HEADER = 0
        val BODY = 1
    }
}